diff --git a/libgcc/config/xtensa/lib1funcs.S b/libgcc/config/xtensa/lib1funcs.S
index b1615ce2..4ce7e555 100644
--- a/libgcc/config/xtensa/lib1funcs.S
+++ b/libgcc/config/xtensa/lib1funcs.S
@@ -840,5 +840,56 @@
 #endif /* L_lshrdi3 */
 
 
+#ifdef L_bswapsi2
+	.align	4
+	.global	__bswapsi2
+	.type	__bswapsi2, @function
+__bswapsi2:
+	leaf_entry sp, 16
+	slli	a4, a2, 8
+	extui	a3, a2, 8, 8
+	or	a3, a3, a4
+	slli	a4, a3, 8
+	extui	a3, a2, 16, 8
+	or	a3, a3, a4
+	slli	a3, a3, 8
+	extui	a2, a2, 24, 8
+	or	a2, a3, a2
+	leaf_return
+	.size	__bswapsi2, . - __bswapsi2
+
+#endif /* L_bswapsi2 */
+
+
+#ifdef L_bswapdi2
+	.align	4
+	.global	__bswapdi2
+	.type	__bswapdi2, @function
+__bswapdi2:
+	leaf_entry sp, 16
+	slli	a4, a3, 8
+	extui	a6, a3, 8, 8
+	or	a6, a6, a4
+	slli	a4, a6, 8
+	extui	a6, a3, 16, 8
+	or	a6, a6, a4
+	extui	a5, a2, 8, 8
+	extui	a4, a3, 24, 8
+	slli	a3, a2, 8
+	or	a5, a5, a3
+	slli	a3, a5, 8
+	extui	a5, a2, 16, 8
+	or	a5, a5, a3
+	slli	a6, a6, 8
+	slli	a5, a5, 8
+	extui	a3, a2, 24, 8
+	or	a3, a5, a3
+	or	a2, a6, a4
+	leaf_return
+	.size	__bswapdi2, . - __bswapdi2
+
+#endif /* L_bswapdi2 */
+
+
 #include "ieee754-df.S"
 #include "ieee754-sf.S"
diff --git a/libgcc/config/xtensa/t-xtensa b/libgcc/config/xtensa/t-xtensa
index fa0aca78..ab914977 100644
--- a/libgcc/config/xtensa/t-xtensa
+++ b/libgcc/config/xtensa/t-xtensa
@@ -10,6 +10,7 @@
 	_fixunsdfsi _fixunsdfdi _floatsidf _floatunsidf \
 	_floatdidf _floatundidf \
 	_truncdfsf2 _extendsfdf2 \
-	_sqrt _recipdf2 _rsqrtdf2
+	_sqrt _recipdf2 _rsqrtdf2 \
+	_bswapsi2 _bswapdi2
 
 LIB2ADD = $(srcdir)/config/xtensa/lib2funcs.S
